src/main/java/io/anserini/rerank/lib/TiebreakerReranker.java

```diff
@@ -0,0 +1,92 @@
+
+/**
+ * Reranker that ensures consistent ordering of documents that have the same score. Scoring ties are broken based on the
+ * lexicographic ordering of the collection docid. This is accomplished by rounding original document scores to the
+ * fourth decimal place, and then adding a tiny score perturbation to break scoring ties.
+ *
+ * This is necessary for repeatable runs: due to multi-threaded indexing, documents are added to the index in arbitrary
+ * order, which makes Lucene's internal mechanism for resolving scoring ties non-deterministic across different indexes.
+ *
+ * Note however, that this reranker also is not sufficient for completely repeatable runs due to scoring ties that span
+ * the rank cutoff <i>k</i>. Due to scoring ties, the top <i>k</i> might vary across indexes; there is nothing that this
+ * reranker can do for such cases. The only solution is to retrieve more than top <i>k</i>, break scoring ties, and then
+ * truncate to tope <i>k</i>.
+ */
+public class TiebreakerReranker implements Reranker {
+    // Sort by score, break ties by lexicographic ordering of collection docid.
+    private class Result implements Comparable<Result> {
+      public float score;
+      public String docid;
+      public int ldocid;
+      public Document doc;
+
+      public Result(float score, int ldocid, String docid, Document doc) {
+        this.score = score;
+        this.ldocid = ldocid;
+        this.docid = docid;
+        this.doc = doc;
+      }
+
+      public int compareTo(Result other) {
+        if (this.score > other.score) {
+          return -1;
+        } else if (this.score < other.score) {
+          return 1;
+        }
+
+        return this.docid.compareTo(other.docid);
+      }
+    }
+
+    @Override
+    public ScoredDocuments rerank(ScoredDocuments docs, RerankerContext context) {
+      SortedSet<Result> sortedResults = new TreeSet<>();
+      for (int i=0; i<docs.documents.length; i++ ) {
+        float rounded = Math.round(docs.scores[i] * 1e4f) / 1e4f;
+        sortedResults.add(new Result(rounded, docs.ids[i],
+            docs.documents[i].getField(FIELD_ID).stringValue(), docs.documents[i]));
+      }
+
+      int numResults = sortedResults.size();
+      ScoredDocuments rerankedDocs = new ScoredDocuments();
+      rerankedDocs.documents = new Document[numResults];
+      rerankedDocs.ids = new int[numResults];
+      rerankedDocs.scores = new float[numResults];
+
+      int i = 0;
+      int dup = 0;
+      float prevScore = 0.0f;
+      for (Result result : sortedResults) {
+        float curScore = result.score;
+
+        // If we encounter ties, we want to perturb the final score a bit.
+        // Note that we can't use equality comparison directly, because in the case of multiple
+        // ties, we would have perturbed the scores, leading the scores to being not equal.
+        if ( prevScore == 0.0f || prevScore - curScore > 1e-4f ) {
```

Comment:
I am not sure I understand this piece of code:
1. Why it is necessary to have the if-else block in the first place? The comparison is already handled by `Result` class, no? The following questions assume it do need further tweak on the scores....
2. Lucene's ranking score has precision of 1e-6 so why using 1e-4 here?
3. For `curScore = curScore - 1e-6f * dup;` If there are more than 2 docs with the same score 1.000001 and the document right below them has a slightly lower score 1.000000, then by lowing the `curScore` a bit at a time `1e-6f * dup` would make some docs scores lower than 1.000000. And this essentially breaks the whole ranking list, no?

Reply:
Added explanation in code.